---
title: 基于依赖的并行化
impact: CRITICAL
impactDescription: 2-10× improvement
tags: 'async, parallelization, dependencies, better-all'
---

## 基于依赖的并行化

对于具有部分依赖关系的操作，使用 `better-all` 来最大化并行性。它会自动在最早的可能时刻启动每个任务。

**Incorrect (profile 不必要地等待 config):**

```typescript
const [user, config] = await Promise.all([
  fetchUser(),
  fetchConfig()
])
const profile = await fetchProfile(user.id)
```

**Correct (config 和 profile 并行运行):**

```typescript
import { all } from 'better-all'

const { user, config, profile } = await all({
  async user() { return fetchUser() },
  async config() { return fetchConfig() },
  async profile() {
    return fetchProfile((await this.$.user).id)
  }
})
```

**无额外依赖的替代方案:**

我们也可以先创建所有 promise，最后执行 `Promise.all()`。

```typescript
const userPromise = fetchUser()
const profilePromise = userPromise.then(user => fetchProfile(user.id))

const [user, config, profile] = await Promise.all([
  userPromise,
  fetchConfig(),
  profilePromise
])
```

Reference: [https://github.com/shuding/better-all](https://github.com/shuding/better-all)
