---
title: 重型组件的动态导入
impact: CRITICAL
impactDescription: directly affects TTI and LCP
tags: 'bundle, dynamic-import, code-splitting, next-dynamic'
---

## 重型组件的动态导入

使用 `next/dynamic` 来懒加载初始渲染时不需要的大型组件。

**Incorrect (Monaco 与主 chunk 一起打包 ~300KB):**

```tsx
import { MonacoEditor } from './monaco-editor'

function CodePanel({ code }: { code: string }) {
  return <MonacoEditor value={code} />
}
```

**Correct (Monaco 按需加载):**

```tsx
import dynamic from 'next/dynamic'

const MonacoEditor = dynamic(
  () => import('./monaco-editor').then(m => m.MonacoEditor),
  { ssr: false }
)

function CodePanel({ code }: { code: string }) {
  return <MonacoEditor value={code} />
}
```
