---
title: 推迟非关键第三方库
impact: MEDIUM
impactDescription: loads after hydration
tags: 'bundle, third-party, analytics, defer'
---

## 推迟非关键第三方库

分析、日志和错误追踪不会阻塞用户交互。应在 hydration 后加载它们。

**Incorrect (阻塞初始 bundle):**

```tsx
import { Analytics } from '@vercel/analytics/react'

export default function RootLayout({ children }) {
  return (
    <html>
      <body>
        {children}
        <Analytics />
      </body>
    </html>
  )
}
```

**Correct (hydration 后加载):**

```tsx
import dynamic from 'next/dynamic'

const Analytics = dynamic(
  () => import('@vercel/analytics/react').then(m => m.Analytics),
  { ssr: false }
)

export default function RootLayout({ children }) {
  return (
    <html>
      <body>
        {children}
        <Analytics />
      </body>
    </html>
  )
}
```
