---
title: 使用被动事件监听器以获得滚动性能
impact: MEDIUM
impactDescription: eliminates scroll delay caused by event listeners
tags: 'client, event-listeners, scrolling, performance, touch, wheel'
---

## 使用被动事件监听器以获得滚动性能

添加 `{ passive: true }` 到触摸和滚轮事件监听器以启用立即滚动。浏览器通常会等待监听器完成以检查是否调用了 `preventDefault()`，从而导致滚动延迟。

**Incorrect:**

```typescript
useEffect(() => {
  const handleTouch = (e: TouchEvent) => console.log(e.touches[0].clientX)
  const handleWheel = (e: WheelEvent) => console.log(e.deltaY)
  
  document.addEventListener('touchstart', handleTouch)
  document.addEventListener('wheel', handleWheel)
  
  return () => {
    document.removeEventListener('touchstart', handleTouch)
    document.removeEventListener('wheel', handleWheel)
  }
}, [])
```

**Correct:**

```typescript
useEffect(() => {
  const handleTouch = (e: TouchEvent) => console.log(e.touches[0].clientX)
  const handleWheel = (e: WheelEvent) => console.log(e.deltaY)
  
  document.addEventListener('touchstart', handleTouch, { passive: true })
  document.addEventListener('wheel', handleWheel, { passive: true })
  
  return () => {
    document.removeEventListener('touchstart', handleTouch)
    document.removeEventListener('wheel', handleWheel)
  }
}, [])
```

**在以下情况使用 passive:** 跟踪/分析、日志记录、任何不调用 `preventDefault()` 的监听器。

**在以下情况不使用 passive:** 实现自定义滑动手势、自定义缩放控制或任何需要 `preventDefault()` 的监听器。
