---
title: 在循环中缓存属性访问
impact: LOW-MEDIUM
impactDescription: reduces lookups
tags: 'javascript, loops, optimization, caching'
---

## 在循环中缓存属性访问

在热路径中缓存对象属性查找。

**Incorrect (3 lookups × N iterations):**

```typescript
for (let i = 0; i < arr.length; i++) {
  process(obj.config.settings.value)
}
```

**Correct (1 total lookup):**

```typescript
const value = obj.config.settings.value
const len = arr.length
for (let i = 0; i < len; i++) {
  process(value)
}
```
