---
title: 提升 RegExp 创建
impact: LOW-MEDIUM
impactDescription: avoids recreation
tags: 'javascript, regexp, optimization, memoization'
---

## 提升 RegExp 创建

不要在 render 内部创建 RegExp。将其提升到模块作用域或使用 `useMemo()` 进行 memoize。

**Incorrect (每次渲染都新建 RegExp):**

```tsx
function Highlighter({ text, query }: Props) {
  const regex = new RegExp(`(${query})`, 'gi')
  const parts = text.split(regex)
  return <>{parts.map((part, i) => ...)}</>
}
```

**Correct (memoize 或提升):**

```tsx
const EMAIL_REGEX = /^[^\s@]+@[^\s@]+\.[^\s@]+$/

function Highlighter({ text, query }: Props) {
  const regex = useMemo(
    () => new RegExp(`(${escapeRegex(query)})`, 'gi'),
    [query]
  )
  const parts = text.split(regex)
  return <>{parts.map((part, i) => ...)}</>
}
```

**Warning (全局 regex 具有可变状态):**

全局 regex (`/g`) 具有可变 `lastIndex` 状态：

```typescript
const regex = /foo/g
regex.test('foo')  // true, lastIndex = 3
regex.test('foo')  // false, lastIndex = 0
```
