---
title: 为重复查找建立索引 Map
impact: LOW-MEDIUM
impactDescription: 1M ops to 2K ops
tags: 'javascript, map, indexing, optimization, performance'
---

## 为重复查找建立索引 Map

同一键的多次 `.find()` 调用应使用 Map。

**Incorrect (每次查找 O(n)):**

```typescript
function processOrders(orders: Order[], users: User[]) {
  return orders.map(order => ({
    ...order,
    user: users.find(u => u.id === order.userId)
  }))
}
```

**Correct (每次查找 O(1)):**

```typescript
function processOrders(orders: Order[], users: User[]) {
  const userById = new Map(users.map(u => [u.id, u]))

  return orders.map(order => ({
    ...order,
    user: userById.get(order.userId)
  }))
}
```

建立 map 一次 (O(n))，然后所有查找均为 O(1)。
对于 1000 个订单 × 1000 个用户：100 万次操作 → 2000 次操作。
