---
title: 提升静态 JSX 元素
impact: LOW
impactDescription: avoids re-creation
tags: 'rendering, jsx, static, optimization'
---

## 提升静态 JSX 元素

将静态 JSX 提取到组件外部以避免重新创建。

**Incorrect (每次渲染都重新创建元素):**

```tsx
function LoadingSkeleton() {
  return <div className="animate-pulse h-20 bg-gray-200" />
}

function Container() {
  return (
    <div>
      {loading && <LoadingSkeleton />}
    </div>
  )
}
```

**Correct (重用相同的元素):**

```tsx
const loadingSkeleton = (
  <div className="animate-pulse h-20 bg-gray-200" />
)

function Container() {
  return (
    <div>
      {loading && loadingSkeleton}
    </div>
  )
}
```

这对大型且静态的 SVG 节点特别有用，因为在每次渲染时重新创建它们可能很昂贵。

**注意:** 如果你的项目启用了 [React Compiler](https://react.dev/learn/react-compiler)，编译器会自动提升静态 JSX 元素并优化组件重渲染，因此不再需要手动提升。
