---
title: 动画化 SVG Wrapper 而不是 SVG 元素
impact: LOW
impactDescription: enables hardware acceleration
tags: 'rendering, svg, css, animation, performance'
---

## 动画化 SVG Wrapper 而不是 SVG 元素

许多浏览器对 SVG 元素上的 CSS3 动画没有硬件加速。将 SVG 包裹在 `<div>` 中并动画化 wrapper。

**Incorrect (直接动画化 SVG - 无硬件加速):**

```tsx
function LoadingSpinner() {
  return (
    <svg 
      className="animate-spin"
      width="24" 
      height="24" 
      viewBox="0 0 24 24"
    >
      <circle cx="12" cy="12" r="10" stroke="currentColor" />
    </svg>
  )
}
```

**Correct (动画化 wrapper div - 硬件加速):**

```tsx
function LoadingSpinner() {
  return (
    <div className="animate-spin">
      <svg 
        width="24" 
        height="24" 
        viewBox="0 0 24 24"
      >
        <circle cx="12" cy="12" r="10" stroke="currentColor" />
      </svg>
    </div>
  )
}
```

这适用于所有 CSS 变换和过渡（`transform`, `opacity`, `translate`, `scale`, `rotate`）。Wrapper div 允许浏览器使用 GPU 加速以获得更流畅的动画。
