---
title: 使用 useRef 存储瞬态值
impact: MEDIUM
impactDescription: avoids unnecessary re-renders on frequent updates
tags: 'rerender, useref, state, performance'
---

## 使用 useRef 存储瞬态值

当一个值频繁更改且你不希望每次更新都重新渲染时（例如，鼠标跟踪器、间隔、瞬态标志），请将其存储在 `useRef` 中而不是 `useState`。保留组件状态用于 UI；使用 refs 用于临时的 DOM 相邻值。更新 ref 不会触发重新渲染。

**Incorrect (每次更新都渲染):**

```tsx
function Tracker() {
  const [lastX, setLastX] = useState(0)

  useEffect(() => {
    const onMove = (e: MouseEvent) => setLastX(e.clientX)
    window.addEventListener('mousemove', onMove)
    return () => window.removeEventListener('mousemove', onMove)
  }, [])

  return (
    <div
      style={{
        position: 'fixed',
        top: 0,
        left: lastX,
        width: 8,
        height: 8,
        background: 'black',
      }}
    />
  )
}
```

**Correct (跟踪时不重新渲染):**

```tsx
function Tracker() {
  const lastXRef = useRef(0)
  const dotRef = useRef<HTMLDivElement>(null)

  useEffect(() => {
    const onMove = (e: MouseEvent) => {
      lastXRef.current = e.clientX
      const node = dotRef.current
      if (node) {
        node.style.transform = `translateX(${e.clientX}px)`
      }
    }
    window.addEventListener('mousemove', onMove)
    return () => window.removeEventListener('mousemove', onMove)
  }, [])

  return (
    <div
      ref={dotRef}
      style={{
        position: 'fixed',
        top: 0,
        left: 0,
        width: 8,
        height: 8,
        background: 'black',
        transform: 'translateX(0px)',
      }}
    />
  )
}
```
