---
title: 将状态读取推迟到使用点
impact: MEDIUM
impactDescription: avoids unnecessary subscriptions
tags: 'rerender, searchParams, localStorage, optimization'
---

## 将状态读取推迟到使用点

如果你只在回调中使用动态状态（searchParams, localStorage），请不要订阅它。

**Incorrect (订阅所有 searchParams 更改):**

```tsx
function ShareButton({ chatId }: { chatId: string }) {
  const searchParams = useSearchParams()

  const handleShare = () => {
    const ref = searchParams.get('ref')
    shareChat(chatId, { ref })
  }

  return <button onClick={handleShare}>Share</button>
}
```

**Correct (按需读取，无订阅):**

```tsx
function ShareButton({ chatId }: { chatId: string }) {
  const handleShare = () => {
    const params = new URLSearchParams(window.location.search)
    const ref = params.get('ref')
    shareChat(chatId, { ref })
  }

  return <button onClick={handleShare}>Share</button>
}
```
